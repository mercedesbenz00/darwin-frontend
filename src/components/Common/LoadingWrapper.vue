<template>
  <div>
    <div
      v-if="loading"
      v-loading="loading"
      :loading-options="options"
      class="spinner"
    />
    <slot v-else />
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'

/**
 * Generic component used to wrap a UI which depends on data being loaded from somewhere.
 * This component appears the same way as v-loading directive
 *
 * The major differences between them (v-loading directive and this component)
 * - v-loading directive renders the main component while loading
 * - This component doesn't render the main component before the loading is finished
 *
 * Properties
 *
 * @property {Boolean} loading
 * - flag indicating if data is currently loading
 * - if true, component will render a loading spinner
 * - if false, component will render the default slot
 *
 * The following properties works same as v-loading directives option params.
 * @property {String} backgroundColor
 * @property {String} label
 * @property {String} size
 * @property {String} theme
 *
 * In an attempt to be a generalized solution, the component is styled to render
 * it's content full width, with the loading indicator having a fixed height of
 * 80px, while the actual content's height grows as needed.
 */
@Component({ name: 'loading-wrapper' })
export default class LoadingWrapper extends Vue {
  @Prop({ required: true, type: Boolean })
  loading!: boolean

  @Prop({ required: false, default: 'rgba(0, 0, 0, 0.1)' })
  backgroundColor!: string

  @Prop({ required: false, default: null })
  label!: string | null

  @Prop({ required: false, default: 'medium' })
  size!: 'small' | 'medium' | 'large'

  @Prop({ required: false, default: 'grey' })
  theme!: 'grey' | 'dark'

  @Prop({ required: false })
  minTimeout?: number

  get options () {
    return {
      backgroundColor: this.backgroundColor,
      label: this.label,
      size: this.size,
      theme: this.theme,
      ...(this.minTimeout && { minTimeout: this.minTimeout })
    }
  }
}
</script>

<style lang="scss" scoped>
.spinner {
  min-width: 30px;
  min-height: 30px;
  height: 100%;
}
</style>
