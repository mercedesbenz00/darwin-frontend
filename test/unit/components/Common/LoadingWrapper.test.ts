import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'

import LoadingWrapper from '@/components/Common/LoadingWrapper.vue'
import loadingDirective from '@/directives/loading'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.directive('loading', loadingDirective)

it('matches snapshot when loading', () => {
  const propsData = { loading: true }
  const slots = { default: '<div>Data slot</div>' }
  const wrapper = shallowMount(LoadingWrapper, { localVue, propsData, slots })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot when loaded', () => {
  const propsData = { loading: false }
  const slots = { default: '<div>Data slot</div>' }
  const wrapper = shallowMount(LoadingWrapper, { localVue, propsData, slots })
  expect(wrapper).toMatchSnapshot()
})
