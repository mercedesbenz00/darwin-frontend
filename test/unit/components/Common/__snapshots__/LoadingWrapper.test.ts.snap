// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot when loaded 1`] = `
<div>
  <div>Data slot</div>
</div>
`;

exports[`matches snapshot when loading 1`] = `
<div>
  <div loading-options="[object Object]" class="spinner" data-min-timeout="200" data-remove="false" data-timeout="false" data-static="true" style="position: relative;">
    <div class="v-loading v-loading--medium v-loading--grey v-loading--active" style="background-color: rgba(0, 0, 0, 0.1);">
      <div class="mul5">
        <div class="mul5circle1"></div>
        <div class="mul5circle2"></div>
        <div class="mul5circle3"></div>
      </div>
    </div>
  </div>
</div>
`;
